### DISCUSSION TITLE
ES8388 board and ESP32

### DISCUSSION THREAD

**Discussion by uutzinger** [pschatzmann/arduino-audio-driver#8] [Q&A] [answered]
I saw a few folks posting about ES8388 having issues with I2C. This issue is real as I can find the device using Arduino Example Wire/WireScan at address 0x10 but audio driver does not communicate on correct I2C settings.

I enabled DEBUG mode in the Example audio-driver/audiotools/audiotools-custom-max and followed the source code. I believe there is an issue with DriverPins.h at

```
  virtual bool begin() {
    AD_LOGD("DriverPins::begin");
...
    // setup i2c
    for (auto &tmp : i2c) {
      result &= tmp.begin();
    }

```
This is not executing 

```
struct PinsI2C {
...
  bool begin() {
    if (set_active) {
      AD_LOGD("PinsI2C::begin for %d", function);
      // if no pins are defined, just call begin
      if (!pinsAvailable()) {
        AD_LOGI("setting up I2C w/o pins");
```

function because **I do not set any debug information regarding I2C SDA, SCL and setclock.**

I added a function to check for the presence of a device on the i2c bus with 
`i2c_bus_check(i2c_bus_handle_t bus, int addr)` which does a `beginTransmission` followed by an `endTransmission` to request an acknowledgement from the device but since the I2C bus is not correctly setup it fails to find the device.

I call this function right at the beginning of es8388_init with:
```
    res = i2c_bus_check(handle, ES8388_ADDR);
    if (res != 0) {
        AD_LOGE("ES8388 not found on I2C bus, check wiring");
        return res;
    } else {
        AD_LOGI("Found ES8388");
    }
```

My output looks like  the following (there is no I2C debug info, that is why I believe the I2C setup is not called):

```
[D] I2SCodecStream.h : 88 - virtual bool audio_tools::I2SCodecStream::begin(audio_tools::I2SCodecConfig)
[D] I2SCodecStream.h : 228 - input: 1
[D] I2SCodecStream.h : 230 - output: 3
[D] I2SCodecStream.h : 245 - BIT_LENGTH_16BITS
[D] I2SCodecStream.h : 259 - RATE_8K
[D] I2SCodecStream.h : 290 - I2S_NORMAL
Debug:   AudioBoard::pins::begin
Debug:   DriverPins::begin
Debug:   DriverPins::begin::SPI
Debug:   DriverPins::begin::I2C

HERE SHOULD I2C debug info printed

Debug:   AudioBoard::pins::begin::returned:true
Debug:   AudioDriver::begin
Debug:   es8388_init

I added this

Debug:   i2c_bus_check: addr=0x10
Error:   ->p_wire->endTransmission: 4
Error:   ES8388 not found on I2C bus, check wiring
Error:   init failed
Debug:   AudioBoard::driver::begin::returned:false
```

I do not think I should add `audio_board.begin()` between these two lines in the example program.

```
AudioBoard                    audio_board(AudioDriverES8388, my_pins); // audio board
I2SCodecStream                i2s_out_stream(audio_board);             // i2s coded
```

I will continue trying to find the issue.

**Comment by pschatzmann** [comment, marked as answer]
Initially the API used ESP32 I2C port naming conventions which is Arduino Port << 1. Since this is confusing, I have changed the logic to use Arduino logic as well, so that the respult from the portscan can be used w/o any changes

https://github.com/pschatzmann/arduino-audio-driver/discussions/6

**Reply by m0rph03nix** [reply]
Thank you. Is there any regression ? I tried the example player-sd-audiokit.ino with kit AudioKitEs8388V2 on a "ESP32 Audio Kit V2.2". All the packages and dependencies are from today. And I have roughly the same problem than @uutzinger one year ago :

[    90][I][esp32-hal-i2c.c:75] i2cInit(): Initialising I2C Master: sda=18 scl=23 freq=100000
Error:   ->p_wire->endTransmission: 2
Error:   ES8388 not found on I2C address 10, check wiring
Error:   AudioDriver init failed
Error:   setConfig has failed
Error:   AudioBoard::driver::begin failed
[E] I2SCodecStream.h : 220 - bool audio_tools::I2SCodecStream::begin1()

Any idea of a recent impactful change ?

**Reply by pschatzmann** [reply]
No change from my side: I was doing some tests with the new ESP32 Arduino core 3.1.2 with an AudioKitEs8388V1 and everything works as expected...
